python-checks/src/main/java/org/sonar/python/checks/cdk/DisabledESDomainEncryptionCheck.java

```diff
@@ -0,0 +1,118 @@
+import org.sonar.plugins.python.api.tree.DictionaryLiteral;
+import org.sonar.plugins.python.api.tree.DictionaryLiteralElement;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.KeyValuePair;
+import org.sonar.plugins.python.api.tree.StringLiteral;
+import org.sonar.plugins.python.api.tree.Tree;
+
+@Rule(key = "S6308")
+public class DisabledESDomainEncryptionCheck extends AbstractCdkResourceCheck {
+  private static final String OMITTING_MESSAGE = "Omitting %s causes encryption of data at rest to be disabled for this %s domain." +
+    " Make sure it is safe here.";
+  private static final String UNENCRYPTED_MESSAGE = "Make sure that using unencrypted %s domains is safe here.";
+  private static final String ENABLED = "enabled";
+  private static final String OPENSEARCH = "OpenSearch";
+  private static final String ELASTICSEARCH = "Elasticsearch";
+
+  @Override
+  protected void registerFqnConsumer() {
+    checkFqn("aws_cdk.aws_opensearchservice.Domain",
+      checkDomain("encryption_at_rest", "aws_cdk.aws_opensearchservice.EncryptionAtRestOptions", OPENSEARCH));
+    checkFqn("aws_cdk.aws_opensearchservice.CfnDomain",
+      checkDomain("encryption_at_rest_options", "aws_cdk.aws_opensearchservice.CfnDomain.EncryptionAtRestOptionsProperty", OPENSEARCH));
+    checkFqn("aws_cdk.aws_elasticsearch.Domain",
+      checkDomain("encryption_at_rest", "aws_cdk.aws_elasticsearch.EncryptionAtRestOptions", ELASTICSEARCH));
+    checkFqn("aws_cdk.aws_elasticsearch.CfnDomain",
+      checkDomain("encryption_at_rest_options", "aws_cdk.aws_elasticsearch.CfnDomain.EncryptionAtRestOptionsProperty", ELASTICSEARCH));
+  }
+
+  private static BiConsumer<SubscriptionContext, CallExpression> checkDomain(String encryptionArgName, String argFqnMethod, String engine) {
+    return (ctx, callExpression) -> getArgument(ctx, callExpression, encryptionArgName)
+      .ifPresentOrElse(
+        argEncryptedTrace -> argEncryptedTrace.addIssueIf(isSensitiveOptionObj(ctx, argFqnMethod).or(isSensitiveOptionDict(ctx)), unencryptedMessage(engine)),
+        () -> ctx.addIssue(callExpression.callee(), omittingMessage(encryptionArgName, engine))
+      );
+  }
+
+  private static  String omittingMessage(String encryptionArgName, String engine) {
+    return String.format(OMITTING_MESSAGE, encryptionArgName, engine);
+  }
+
+  private static String unencryptedMessage(String engine) {
+    return String.format(UNENCRYPTED_MESSAGE, engine);
+  }
+
+  private static Predicate<Expression> isSensitiveOptionObj(SubscriptionContext ctx, String argFqnMethod) {
+    return expr -> {
+      if (!isFqn(argFqnMethod).test(expr)) {
+        return false;
+      }
+      if (!expr.is(Tree.Kind.CALL_EXPR)) {
+        return true;
+      }
+
+      return getArgument(ctx, (CallExpression) expr, ENABLED)
+        .filter(argEnabledTrace -> !argEnabledTrace.hasExpression(isFalse())).isEmpty();
+    };
```

Comment:
Works, but can be improved to have more findings.
What if the function call, which is actually an object initiation, takes place outside the call and only a variable is used here? I would honestly only check the initiation of the option and not check if a function call is in a function call.
Meaning any call to `aws_cdk.aws_elasticsearch.EncryptionAtRestOptions` as an example. It may be that an option object is not used. But by definition, it is already insecure and should not be used in a domain. WDYT?

Reply:
Mmhh not totally sure what you mean here.
If I understand correctly, you are talking about detecting such usecase :
```
encryptionOption = aws_os.EncryptionAtRestOptions(enabled=False)
aws_os.Domain(encryption_at_rest=encryptionOption) # NonCompliant
```

This usecase is already detected because this method `isSensitiveOptionObj` is called on each expressions of the argumentTrace of the argument `encryption_at_rest`.

Your approach seems easier and possibly more reliable, but I see two relative disadvantage :
- as you say we will raise an issue even if the option object is not used (not a big issue from my perspective)
- we would not be able to easily highlight the call to `Domain` which would use this variable, although this does not seems to be specified in the RSPEC (no highlight.doc ?)

Please let me know if you meant something else.

Reply:
You're right. Missed the point, that every expression is checked!